File: string/reverse-words-in-a-string-iii/reverse-words-in-a-string-iii.ts
function reverseWords(s: string): string {
    const reversed = Array<string>(s.length)

    let j = 0;
    for (let i = 0; i < s.length; i++) {
        if (s[i] === " ") {
            reverseWord(s.slice(j, i), reversed, j)
            j = i + 1;
        }
    }

    return reversed.join(" ")
};

function reverseWord(s: string, arr: string[], start: number) {
    let j = start + s.length - 1;
    for(let i = 0; i < s.length; i++) {
        arr[j--] = s[i];
    }
}

File: string/reverse-words-in-a-string-iii/reverse-words-in-a-string-iii.go
package reversewordsinastringiii

func reverseWords(s string) string {
	reversed := []rune(s)

	b := 0
	var e int
	for e = 0; e < len(reversed); e++ {
		if reversed[e] == ' ' {
			reverse(reversed, b, e-1)
			b = e + 1
		}
	}
	reverse(reversed, b, e-1)

	return string(reversed)
}

func reverse(s []rune, b, e int) {
	j := e
	i := b
	for i < j {
		tmp := s[i]
		s[i] = s[j]
		s[j] = tmp
		j--
		i++
	}
}
